webapp/sources/ldap-inventory/inventory-provisioning-web/src/main/scala/com/normation/inventory/provisioning/endpoint/InventoryProcessor.scala

```diff
@@ -114,123 +110,122 @@ class InventoryProcessor(
+  /*
+   * We manage inventories buffering with a bounded queue.
+   * In case of overflow, we reject new inventories.
+   */
+  val queue = ZioRuntime.unsafeRun(Queue.dropping[InventoryReport](maxQueueSize))
 
-  // the asynchrone, bufferised processor
-  lazy val reportProcess = {
-    // the synchronize report processor
-    // we use the i/o scheduler given the kind of task for report processor
-    val syncReportProcess = Subscriber(new ProcessInventoryObserver(queueSize), Scheduler.io())
+  def currentQueueSize: Int = ZioRuntime.unsafeRun(queue.size)
 
-    // and we async/buf it. The "overflow" strategy is unbound, because we
-    // will manage it by hand with the queuesize
-    BufferedSubscriber[InventoryReport](syncReportProcess, Unbounded)
+  /*
+   * Saving reports is a loop which consume items from queue
+   */
+  def loop(): UIO[Nothing] = {
+    queue.take.flatMap(saveReport).flatMap(_ => loop())
   }
 
+  // start the inventory processing
+  ZioRuntime.unsafeRun(loop().fork)
+
   // we need something that can produce the stream several time because we are reading the file several time - not very
   // optimized :/
-  def saveInventory(newInventoryStream: () => InputStream, inventoryFileName: String, optNewSignatureStream : Option[() => InputStream]): Box[InventoryProcessStatus] = {
-
-    def saveWithSignature(report: InventoryReport, newSignature: () => InputStream): Box[InventoryProcessStatus] = {
-      val signatureStream = newSignature()
-      val inventoryStream = newInventoryStream()
-      val response = for {
-        digest  <- digestService.parse(signatureStream)
-        (key,_) <- digestService.getKey(report)
-        checked <- digestService.check(key, digest, inventoryStream)
-        saved   <- if (checked) {
-                     // Signature is valid, send it to save engine
-                     // Set the keyStatus to Certified
-                     // For now we set the status to certified since we want pending inventories to have their inventory signed
-                     // When we will have a 'pending' status for keys we should set that value instead of certified
-                     val certifiedReport = report.copy(node = report.node.copyWithMain(main => main.copy(keyStatus = CertifiedKey)))
-                     checkQueueAndSave(certifiedReport)
-                   } else {
-                     // Signature is not valid, reject inventory
-                     Full(InventoryProcessStatus.SignatureInvalid(report))
-                   }
-       } yield {
-        saved
+  def saveInventory(info: SaveInventoryInfo): IOResult[InventoryProcessStatus] = {
+
+    def saveWithSignature(report: InventoryReport, newInventoryStream: () => InputStream, newSignature: () => InputStream): IOResult[InventoryProcessStatus] = {
+      ZIO.bracket(Task.effect(newInventoryStream()).mapError(SystemError(s"Error when reading inventory file '${report.name}'", _)))(is => Task.effect(is.close()).run) { inventoryStream =>
```

Comment:
brackets are love

Reply:
Your bracket is open, can you zip it ? (just joking with the french word "braguette", ignore that comment, please)